pkg/distributor/distributor.go

```diff
@@ -448,6 +454,28 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	var validationErrors util.GroupedErrors
 	validationContext := d.validator.getValidationContextForTime(time.Now(), tenantID)
 
+	shardStreamsCfg := d.validator.Limits.ShardStreams(tenantID)
+	maybeShardByRate := func(stream logproto.Stream, pushSize int) {
+		if shardStreamsCfg.Enabled {
+			streams = append(streams, d.shardStream(stream, pushSize, tenantID)...)
+		} else {
+			streams = append(streams, KeyedStream{
+				HashKey: lokiring.TokenFor(tenantID, stream.Labels),
+				Stream:  stream,
+			})
+		}
+	}
+	maybeShardByTime := func(stream logproto.Stream, labels labels.Labels, pushSize int) {
+		if shardStreamsCfg.TimeShardingEnabled {
+			streamsByTime := shardStreamByTime(stream, labels, d.ingesterCfg.MaxChunkAge/2)
+			for _, ts := range streamsByTime {
+				maybeShardByRate(ts.Stream, ts.linesTotalLen)
+			}
+		} else {
+			maybeShardByRate(stream, pushSize)
+		}
+	}
```

Comment:
This code can be moved out to a separate function not to overwhelm Push function and making it more clear, because right now it's confusing a bit that we call the function to shard by time but for the most of the users it will shard by rate. 
Nothing critical, just think that it will be better to have it separately ;)

Reply:
I played a bit with this, but it would require quite a bit of refactoring to do that cleanly, or a whole bunch of parameters for every method, so I chose to leave if for the future :sweat_smile:

Reply:
sure, np